snap/squashfs/squashfs.go

```diff
@@ -74,32 +76,66 @@ func (s *Snap) Install(targetPath, mountDir string) error {
 	return runCommand("cp", "-a", s.path, targetPath)
 }
 
-var runCommand = func(args ...string) error {
+var runCommandWithOutput = func(args ...string) ([]byte, error) {
 	cmd := exec.Command(args[0], args[1:]...)
-	if output, err := cmd.CombinedOutput(); err != nil {
-		return fmt.Errorf("cmd: %q failed: %v (%q)", strings.Join(args, " "), err, output)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("cmd: %q failed: %v (%q)", strings.Join(args, " "), err, output)
 	}
-	return nil
+
+	return output, nil
+}
+
+var runCommand = func(args ...string) error {
+	_, err := runCommandWithOutput(args...)
+	return err
 }
 
 func (s *Snap) unpack(src, dstDir string) error {
 	return runCommand("unsquashfs", "-f", "-i", "-d", dstDir, s.path, src)
 }
 
 // ReadFile returns the content of a single file inside a squashfs snap.
-func (s *Snap) ReadFile(path string) (content []byte, err error) {
+func (s *Snap) ReadFile(filePath string) (content []byte, err error) {
 	tmpdir, err := ioutil.TempDir("", "read-file")
 	if err != nil {
 		return nil, err
 	}
 	defer os.RemoveAll(tmpdir)
 
 	unpackDir := filepath.Join(tmpdir, "unpack")
-	if err := runCommand("unsquashfs", "-i", "-d", unpackDir, s.path, path); err != nil {
+	if err := runCommand("unsquashfs", "-i", "-d", unpackDir, s.path, filePath); err != nil {
 		return nil, err
 	}
 
-	return ioutil.ReadFile(filepath.Join(unpackDir, path))
+	return ioutil.ReadFile(filepath.Join(unpackDir, filePath))
+}
+
+// ListDir returns the content of a single directory inside a squashfs snap.
+func (s *Snap) ListDir(dirPath string) ([]string, error) {
+	output, err := runCommandWithOutput(
+		"unsquashfs", "-no-progress", "-dest", "", "-l", s.path, dirPath)
+	if err != nil {
+		return nil, err
+	}
+
+	dirPath = path.Clean(dirPath)
+	pattern := regexp.MustCompile(regexp.QuoteMeta(dirPath) + string(os.PathSeparator) + "?(.+)")
```

Comment:
MustCompile is strange given that dirPath comes from outside, also I would add start/endline anchoring otherwise there may be strange corner cases, also if you exclude /  ([^/] instead of .) from the tail end you don't need the check ContainsRune check?